packages/ics721/src/ibc_packet_receive.rs

```diff
@@ -62,25 +31,83 @@ pub(crate) fn receive_ibc_packet(
     let data: NonFungibleTokenPacketData = from_json(&packet.data)?;
     data.validate()?;
 
-    let cloned_data = data.clone();
-    let receiver = deps.api.addr_validate(&data.receiver)?;
-    let token_count = data.token_ids.len();
-
     // Check if NFT is local if not get the local class id
     let maybe_local_class_id = try_pop_source_prefix(&packet.src, &data.class_id);
     let callback = get_receive_callback(&data);
+    let local_class_id = if let Some(local_class_id) = maybe_local_class_id {
+        ClassId::new(local_class_id)
+    } else {
+        let local_prefix = get_endpoint_prefix(&packet.dest);
+        ClassId::new(format!("{}{}", local_prefix, data.class_id))
+    };
 
-    let action_aggregator = data
+    // sub message holds 2 to 4 messages:
+    // - one message for voucher creation or redemption, another message for updating incoming or outgoing channel
+    let (is_redemption, voucher_and_channel_messages) = create_voucher_and_channel_messages(
+        deps.as_ref(),
+        env.clone(),
+        data.clone(),
+        maybe_local_class_id,
+        local_class_id.clone(),
+        packet.clone(),
+    )?;
```

Comment:
new logic leads to same result, by creating required messages - without using temporary action and aggregator objects